src/world.cpp

```diff
@@ -2208,7 +2208,7 @@ void World_GenRoom(struct room_s *room, class VT_Level *tr)
     room->alternate_room = NULL;
     room->base_room = NULL;
 
-    if((tr_room->alternate_room >= 0) && ((uint32_t)tr_room->alternate_room < tr->rooms_count))
+    if((tr_room->alternate_room >= 0) && ((uint32_t)tr_room->alternate_room < tr->rooms_count) && (room->id < tr_room->alternate_room))
```

Comment:
As said, TR1 swaps the _contents_ of the room structure, not pointers. It does not use any "base room" reference or such; it only swaps the contents (i.e., all items that can have a room ownership) using the alternative room id; if it is `<0`, no swapping occurs.

Example: If room 0 has alternative=1, and room 1 has alternative=2, (i.e. [1,2,...]) on the first swap all items are removed from room 0 and moved to room 1 (discarding all items from room 1), then the contents of the structures are swapped, so the data occurs as if room 1 has been the "base room" from the beginning. After the swap, the alternative ids are now [2, -1,...], because the alternative room's alternative room id (yikes!) is set to -1.